### PR TITLE
[ JPG - 007] Create Transaction: DEBIT

### PR DESCRIPTION
**Pull request by Ketan-SP18**
-Add Debit API

### CODE REVIEW

File: src/main/java/com/passbook/sparkeighteen/controller/TransactionController.java [hunk 1]
```diff
@@ -0,0 +1,39 @@
...
+@RequestMapping("/api/v1/user")
+@RestController
+public class TransactionController {
+    private final TransactionService transactionService;
+
+    public TransactionController(TransactionService transactionService) {
+        this.transactionService = transactionService;
+    }
+
+    @ApiOperation("User deposits a certain amount")
```
Comment (spark-coders) [inline src/main/java/com/passbook/sparkeighteen/controller/TransactionController.java:27]: Why does your code has this?

File: src/main/java/com/passbook/sparkeighteen/service/UserService.java [hunk 2]
```diff
@@ -54,6 +54,7 @@ public LoginResponse login(final LoginRequest userLogin) throws Exception {
 
         return LoginResponse.builder()
                 .userID(user.getId())
+                .firstname(user.getFirstname())
```
Comment (spark-coders) [inline src/main/java/com/passbook/sparkeighteen/service/UserService.java:57]: Is it part of your PR?

File: src/main/resources/application.properties [hunk 3]
```diff
@@ -1,7 +1,7 @@
 spring.datasource.url=jdbc:postgresql://localhost:5432/passbook
 spring.datasource.username=postgres
 spring.datasource.password=qwerty1201
-#spring.jpa.hibernate.ddl-auto=create-drop
+#spring.jpa.hibernate.ddl-auto=create
```
Comment (spark-coders) [inline src/main/resources/application.properties:4]: Same mistake again and again

File: src/main/java/com/passbook/sparkeighteen/service/TransactionService.java [hunk 4]
```diff
@@ -0,0 +1,120 @@
...
+                .user(user)
+                .closingBalance(request.getAmount() + getZeroOrLastBalance(user))
+                .time(LocalDateTime.now())
+                .transactionType(TransactionType.CREDIT)
+                .build());
+
+        return Optional.of(transaction);
+    }
+
+    private Float getZeroOrLastBalance(UserEntity user) {
```
Comment (spark-coders) [inline src/main/java/com/passbook/sparkeighteen/service/TransactionService.java:107]: Why is method?

File: src/main/java/com/passbook/sparkeighteen/service/TransactionService.java [hunk 5]
```diff
@@ -0,0 +1,120 @@
...
+                .amount(transaction.getAmount())
+                .note(transaction.getNote())
+                .time(transaction.getTime())
+                .closingBalance(transaction.getClosingBalance())
+                .transactionType(transaction.getTransactionType())
+                .message(String.format("Your A/C XXXXX has a %s by Rs %.2f on %s ", transaction.getTransactionType(), transaction.getAmount(), transaction.getTime()))
+                .build();
+    }
+
+    private Optional<TransactionEntity> withdraw(UserEntity user, TransactionRequest request) {
```
Comment (spark-coders) [inline src/main/java/com/passbook/sparkeighteen/service/TransactionService.java:70]: Over complicated code think about refactoring.

File: src/main/java/com/passbook/sparkeighteen/service/TransactionService.java [hunk 6]
```diff
@@ -0,0 +1,120 @@
...
+                .user(user)
+                .closingBalance(getZeroOrLastBalance(user) - request.getAmount())
+                .time(LocalDateTime.now())
+                .transactionType(TransactionType.DEBIT)
+                .build());
+
+        return Optional.of(transaction);
+    }
+
+    private Optional<TransactionEntity> deposit(UserEntity user, TransactionRequest request) {
```
Comment (spark-coders) [inline src/main/java/com/passbook/sparkeighteen/service/TransactionService.java:90]: Your PR should not have this code.

File: src/main/java/com/passbook/sparkeighteen/service/TransactionService.java [hunk 7]
```diff
@@ -0,0 +1,120 @@
...
+import com.passbook.sparkeighteen.peristence.repository.TransactionRepository;
+import com.passbook.sparkeighteen.peristence.repository.UserRepository;
+import org.springframework.stereotype.Service;
+
+import java.time.LocalDateTime;
+import java.util.List;
+import java.util.Optional;
+
+@Service
+public class TransactionService {
```
Comment (spark-coders) [inline src/main/java/com/passbook/sparkeighteen/service/TransactionService.java:17]: Over complicated logic

**Review by spark-coders — COMMENTED** [review]
Test case missing.
Code is common from 006.
Too much code is common and business logic is too complicated need to think about the a better approch.

**Comment by Ketan-SP18** [comment]
Closed this, as will refactor the Deposit api with Credit api in a single api

**Pull request closed without merging**